[Virtual Point Clouds] Allow editing of files that make up the VPC

ViperMiniQ opened the pull request:
This PR enables the editing of individual point cloud datasets that make up the virtual point cloud.
Multiple point clouds are editable this way in one go.

All or none, we do not allow editing of only a portion of the VPC, all files have to be editable for the editing tool to allow making changes to the layer. In other words, if a sub index of the VPC is unreachable in any way (deleted, corrupted, not supported...), the whole layer is locked for editing.

For a VPC to be editable, it needs to reference COPC files only and all files have to be local.

ViperMiniQ committed:
expose uri of the point cloud index

ViperMiniQ committed:
create editing index from uri

ViperMiniQ committed:
fix docs

ViperMiniQ committed:
silence clang tidy

ViperMiniQ committed:
fix docs

ViperMiniQ committed:
remove uri from the constructor

ViperMiniQ committed:
remove unused variable

ViperMiniQ committed:
prepare selected points in the 3d map, adjust function calls

ViperMiniQ committed:
add support for vpc editing in pointcloudlayer

undo stack has been modified to track the uri of the changed point cloud index, function calls have been modified to allow that

where appropriate, a clear distinction is kept between the vpc and simple point cloud layer

ViperMiniQ committed:
sipify

ViperMiniQ committed:
fix crashes when calling uri method on nullptr indexes

also, fix formatting

ViperMiniQ committed:
sipify

ViperMiniQ committed:
remove unused vars

uclaros commented on src/3d/qgspointcloudlayerchunkloader_p.cpp:
```diff
@@ -299,7 +299,11 @@ QgsPointCloudLayerChunkedEntity::QgsPointCloudLayerChunkedEntity( Qgs3DMapSettin
     mChunkUpdaterFactory = std::make_unique<QgsChunkUpdaterFactory>( mChunkLoaderFactory );
 
-    connect( pcl, &QgsPointCloudLayer::chunkAttributeValuesChanged, this, [this]( const QgsPointCloudNodeId &n ) {
+    connect( pcl, &QgsPointCloudLayer::chunkAttributeValuesChanged, this, [this]( const QString uri, const QgsPointCloudNodeId &n ) {
```
```suggestion
    connect( pcl, &QgsPointCloudLayer::chunkAttributeValuesChanged, this, [this]( const QString &uri, const QgsPointCloudNodeId &n ) {
```

uclaros commented on src/app/3d/qgs3dmaptoolpointcloudchangeattribute.cpp:
```diff
@@ -103,12 +131,13 @@ SelectedPoints Qgs3DMapToolPointCloudChangeAttribute::searchPoints( QgsPointClou
   {
     QgsEventTracing::ScopedEvent _trace( u"PointCloud"_s, u"Qgs3DMapToolPointCloudChangeAttribute::searchPoints, looking for affected nodes"_s );
     QgsPointCloudIndex index = layer->index();
```
This is no longer used, so we can name the `pc` parameter to `index`
```suggestion
```

uclaros commented on src/core/pointcloud/qgspointcloudlayer.cpp:
```diff
@@ -1001,41 +1003,113 @@ void QgsPointCloudLayer::loadIndexesForRenderContext( QgsRenderContext &renderer
   {
-    mEditIndex = QgsPointCloudIndex();
-    return false;
+    QgsVirtualPointCloudProvider *vpcProvider = dynamic_cast<QgsVirtualPointCloudProvider *>( mDataProvider.get() );
```
```suggestion
    QgsVirtualPointCloudProvider *vpcProvider = qobject_cast<QgsVirtualPointCloudProvider *>( mDataProvider.get() );
```

uclaros commented on src/core/pointcloud/qgspointcloudlayer.cpp:
```diff
@@ -1051,26 +1125,60 @@ QString QgsPointCloudLayer::commitError() const
+
+      // emitting layerModified() is not required as that's done automatically
+      // when undo stack index changes
+    }
```
empty check is redundant
```suggestion
    for ( const QgsPointCloudNodeId &n : std::as_const( updatedNodes ) )
      emit chunkAttributeValuesChanged( uri, n );

    // emitting layerModified() is not required as that's done automatically
    // when undo stack index changes
```

uclaros commented on src/core/pointcloud/qgspointcloudlayer.cpp:
```diff
@@ -1082,97 +1190,181 @@ bool QgsPointCloudLayer::supportsEditing() const
-  if ( !mEditIndex )
+  if ( mIsVpc )
+  {
+    for ( const QgsPointCloudIndex &index : mEditingSubIndexes )
```
```suggestion
    for ( const QgsPointCloudIndex &index : std::as_const( mEditingSubIndexes ) )
```

ViperMiniQ committed:
address review

remove unused method
use as_const for qvector iteration

ViperMiniQ committed:
expose vpc methods, store sub index position in the undo buffer

- in chunkedentity, use -1 for singular point cloud, -2 for overview, 0 - n for referencing subIndex of a VPC (same for emitting the attribute changed signal)

- return subindexes on request, modified to point to editIndex if the layer is editable and the underlaying index has been loaded

ViperMiniQ committed:
remove uri access

ViperMiniQ committed:
remove uri

ViperMiniQ committed:
re-expose uri for editindex constructor

ViperMiniQ committed:
add static method to decide index loading, fix autos and consts

wonder-sk commented on src/core/pointcloud/qgspointcloudlayer.h:
```diff
@@ -325,17 +326,17 @@ class CORE_EXPORT QgsPointCloudLayer : public QgsMapLayer, public QgsAbstractPro
      * Attempts to modify attribute values for specific points in the editing buffer.
      *
-     * \param nodesAndPoints A list of nodes to modify and points to modify within those nodes
+     * \param mappedPoints A map of URIs to nodes to modify and points to modify within those nodes
```
This needs fixing - we're not using URIs... with a better explanation of the "int" key, and what its value should be in both cases (using VPC or not).

wonder-sk reviewed (commented):
Some more comments 😅

ViperMiniQ committed:
address review

- use QHash instead of QMap for mapping points to subindexes
- return QgsPointCloudSubIndex instead of QgsPointCloudIndex in layer::subIndex() method
- add mEditable to check the editing state of the point cloud layer (false if editing hasn't started, true if it has)
- add default QgsPointCloudSubIndex constructor
- fix docs

ViperMiniQ committed:
add tests

ViperMiniQ commented on src/core/pointcloud/qgspointcloudeditingindex.cpp:
```diff
@@ -48,6 +48,25 @@ QgsPointCloudEditingIndex::QgsPointCloudEditingIndex( QgsPointCloudLayer *layer
+  mRootBounds = mIndex.rootNodeBounds();
+  mSpan = mIndex.span();
+  mIsValid = true;
+}
```
Original constructor removed, kept only the new one.

ViperMiniQ committed:
remove layer::subIndex, use layer::subIndexes everywhere

ViperMiniQ committed:
fix seg fault calling non existing index

ViperMiniQ committed:
editing only available if all local files and all copc

ViperMiniQ committed:
warn user if they start editing on a not completely loaded VPC

ViperMiniQ committed:
remove default construct for subindex

ViperMiniQ committed:
fix variable redefinition

ViperMiniQ committed:
fun test

ViperMiniQ committed:
local copy
